### PR TITLE
GTA IV VRAM fix

### PR DESCRIPTION
**Pull request by K0bin**
As discussed on Discord.

### CODE REVIEW

**Comment by Blisto91** [comment]
Remember example config

**Comment by K0bin** [comment]
Done, thanks for the reminder.

**Review by Blisto91 — APPROVED** [review]

**Review by doitsujin — APPROVED** [review]

**Comment by emansom** [comment]
What was the reasoning behind this?

As far as i'm aware the most recent version available in Steam supports high VRAM amounts just fine?
If not, there is also [launch options](https://gillian-guide.github.io/additional-setup/#launch-options) to limit the VRAM amount that should/can be used instead?

**Comment by emansom** [comment]
@ThirteenAG any insight/recommendations on what DXVK should do for GTA IV to workaround these problems, if at all? and are these bugs even still present?

**Comment by Blisto91** [comment]
Due to lack of some NVAPI or AGS/atiadlxx support the game will think we are on Intel and will behave weirdly with the max amount of vram allowed (it will restrict it to 512MB) when high amounts of dedicated memory is exposed. Telling the game we have a low amount of dedicated memory but high amount of shared makes it happy.

**Comment by emansom** [comment]
> Due to lack of some NVAPI or AGS/atiadlxx support the game will think we are on Intel and will behave weirdly with the max amount of vram allowed (it will restrict it to 512MB) when high amounts of dedicated memory is exposed. Telling the game we have a low amount of dedicated memory but high amount of shared makes it happy.

I'd like to test this claim of "will behave weirdly with the max amount of vram allowed" as it seems to be birthed in ignorance of previous established RAGE engine quirks and workarounds heavily documented in the GTA IV modding community.

To be clear, so I can ask other people to utilize the launcher option vram limiting (see previous post link) instead of DXVK doing that job: 

For testing, do entries in `dxvk.conf` take priority over built-in options or vice-versa?

Is the config value lookup priority as following?:
1. Environment variables
2. `dxvk.conf`
3. built-in overrides in `config.cpp`

Or is the config value lookup priority this?: 
1. built-in overrides in `config.cpp`
2. Environment variables
3. `dxvk.conf`

In other words, do testers need to recompile DXVK to test the validity of the need for this `maxDeviceMemory` override, or is a `dxvk.conf` override to e.g. `3072` enough?

If it is the latter, i'm currently testing `dxgi. maxDeviceMemory = 3072` and so far its a pleasant stutter free experience.

**Comment by Blisto91** [comment]
I just tested it and yes it is still needed for the game to work out of the box with dxvk on Linux. Else the graphical settings will be restricted by the game thinking only 512MB vram is avaliable.
The `dxgi.maxDeviceMemoryconfig` doesn't do anything on Windows since the game loads dxgi from the system folder. If that is where you are testing.

> Is the config value lookup priority as following?:
>
>   Environment variables
>    dxvk.conf
>    built-in overrides in config.cpp

Yes this is true. No need to recompile

**Comment by emansom** [comment]
> I just tested it and yes it is still needed for the game to work out of the box with dxvk on Linux. Else the graphical settings will be restricted by the game thinking only 512MB vram is avaliable.

Yes this is a known problem within the GTA
IV modding community that is already heavily documented and workarounds have already been established, which I pointed out in my previous comment. It seems this whole PR was made in ignorance of it.

Test with latest Steam version of the game, with the following Steam game launch options (right click the game in Steam Library then Properties) as documented [here](https://gillian-guide.github.io/resources/troubleshooting/) and [here](https://www.pcgamingwiki.com/wiki/Grand_Theft_Auto_IV#VRAM_amount_is_wrongly_detected).

`-norestrictions -nomemrestrict -availablevidmem 3072`

It should report VRAM as 3072MB with those options.

> The `dxgi.maxDeviceMemoryconfig` doesn't do anything on Windows since the game loads dxgi from the system folder. If that is where you are testing.

Noted. Yes, Windows.

> Yes this is true. No need to recompile

Great!

**Comment by Blisto91** [comment]
We are more concerned with it working out of the box like native Windows without any tweaks or commands needed.

**Comment by emansom** [comment]
> We are more concerned with it working out of the box like native Windows without any tweaks or commands needed.

GTA IV is one of those games that require mods to have a [pleasant experience](https://www.youtube.com/watch?v=UuXVYUGJ45Y)

It's Rockstar/TakeTwo's task to resolve the out of the box experience ultimately, and they won't. So it's up to the modding community that did a great job with FusionFix.

Did the VRAM show up as 3072MB on Linux with the instructions above?

**Comment by Blisto91** [comment]
It shows up as 3072MB without on Linux because of the builtin dxvk config.

Edit: I am aware of such launch options being needed with dxvk on Windows. They work there yes.

**Comment by K0bin** [comment]
Is there a genuine problem here? Because as far as I know the game works fine.

> GTA IV is one of those games that require mods to have a [pleasant experience](https://www.youtube.com/watch?v=UuXVYUGJ45Y)

The game is just shit and no amount of mods will fix that.
